File: sdk/python/kucoin_universal_sdk/internal/infra/default_ws_callback.py
from threading import Lock, RLock
from typing import Dict, List

from ..interfaces.websocket import WebSocketMessageCallback
from ..util.sub import SubInfo


class Callback:
    def __init__(self, callback: WebSocketMessageCallback, id: str, topic: str):
        self.callback = callback
        self.id = id
        self.topic = topic


class CallbackManager:
    def __init__(self, topic_prefix: str):
        self.id_topic_mapping: Dict[str, Dict[str, bool]] = {}
        self.topic_callback_mapping: Dict[str, Callback] = {}
        self.lock = RLock()
        self.topic_prefix = topic_prefix

    def is_empty(self) -> bool:
        with self.lock:
            return len(self.id_topic_mapping) == 0 and len(self.topic_callback_mapping) == 0

    def get_sub_info(self) -> List[SubInfo]:
        with self.lock:
            sub_info_list = []
            for topics in self.id_topic_mapping.values():
                info = SubInfo(prefix=self.topic_prefix, args=[], callback=None)
                for topic in topics:
                    parts = topic.split(":")
                    if len(parts) == 2:
                        info.args.append(parts[1])
                    if topic in self.topic_callback_mapping:
                        info.callback = self.topic_callback_mapping[topic].callback
                sub_info_list.append(info)
            return sub_info_list

    def add(self, sub_info: SubInfo) -> bool:
        with self.lock:
            id = sub_info.to_id()

            if id in self.id_topic_mapping:
                return False

            topic_map = {}

            for topic in sub_info.topics():
                if topic in self.topic_callback_mapping:
                    continue

                topic_map[topic] = True
                self.topic_callback_mapping[topic] = Callback(
                    callback=sub_info.callback,
                    id=id,
                    topic=topic
                )

            self.id_topic_mapping[id] = topic_map
            return True

    def remove(self, id: str) -> None:
        with self.lock:
            topic_map = self.id_topic_mapping.get(id)
            if not topic_map:
                return

            del self.id_topic_mapping[id]
            for topic in topic_map:
                self.topic_callback_mapping.pop(topic, None)

    def get(self, topic: str) -> WebSocketMessageCallback:
        with self.lock:
            cb = self.topic_callback_mapping.get(topic)
            if cb:
                return cb.callback


class TopicManager:
    def __init__(self):
        self.topic_prefix = {}
        self.lock = Lock()

    def get_callback_manager(self, topic: str) -> CallbackManager:
        parts = topic.split(':')
        prefix = topic
        if len(parts) == 2 and parts[1] != "all":
            prefix = parts[0]

        with self.lock:
            if prefix not in self.topic_prefix:
                self.topic_prefix[prefix] = CallbackManager(topic)
            return self.topic_prefix[prefix]

    def range(self, func):
        with self.lock:
            for key, value in self.topic_prefix.items():
                if not func(key, value):
                    break

File: sdk/golang/internal/infra/default_ws_callback.go
package infra

import (
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/internal/interfaces"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/internal/util"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/common/logger"
	"strings"
	"sync"
)

type Callback struct {
	callback interfaces.WebSocketMessageCallback // callback
	id       string                              // sub id
	topic    string                              // real topic
}

type CallbackManager struct {
	idTopicMapping       map[string]map[string]bool // id : topics
	topicCallbackMapping map[string]*Callback       // topic : callback
	lock                 sync.RWMutex               // lock
	topicPrefix          string                     // topic prefix for pattern matching
}

func NewCallbackManager(topicPrefix string) *CallbackManager {
	return &CallbackManager{
		idTopicMapping:       map[string]map[string]bool{},
		topicCallbackMapping: map[string]*Callback{},
		lock:                 sync.RWMutex{},
		topicPrefix:          topicPrefix,
	}
}

func (s *CallbackManager) Empty() bool {
	s.lock.Lock()
	defer s.lock.Unlock()
	return len(s.idTopicMapping) == 0 && len(s.topicCallbackMapping) == 0
}

func (s *CallbackManager) GetSubInfo() []*util.SubInfo {
	s.lock.Lock()
	defer s.lock.Unlock()

	subInfo := make([]*util.SubInfo, 0)
	for _, topics := range s.idTopicMapping {
		info := &util.SubInfo{
			Prefix: s.topicPrefix,
			Args:   make([]string, 0),
		}
		for topic, _ := range topics {
			parts := strings.Split(topic, ":")
			if len(parts) == 2 {
				info.Args = append(info.Args, parts[1])
			}
			info.Callback = s.topicCallbackMapping[topic].callback
		}

		subInfo = append(subInfo, info)
	}
	return subInfo
}

func (s *CallbackManager) Add(subInfo *util.SubInfo) (created bool) {
	s.lock.Lock()
	defer s.lock.Unlock()

	id := subInfo.ToId()

	_, exist := s.idTopicMapping[id]
	if exist {
		logger.GetLogger().Tracef("callback already exist, id:[%s]", id)
		return false
	}

	topicMap := make(map[string]bool)

	for _, topic := range subInfo.Topics() {

		_, exist := s.topicCallbackMapping[topic]
		if exist {
			logger.GetLogger().Tracef("topic already exist, topic:[%v]", topic)
			continue
		}

		topicMap[topic] = true

		s.topicCallbackMapping[topic] = &Callback{
			callback: subInfo.Callback,
			id:       id,
			topic:    topic,
		}
	}

	s.idTopicMapping[id] = topicMap
	logger.GetLogger().Tracef("id:[%s], topicMap:[%v]", id, topicMap)
	return true
}

func (s *CallbackManager) Remove(id string) {
	s.lock.Lock()
	defer s.lock.Unlock()

	topicMap, exist := s.idTopicMapping[id]
	if !exist {
		logger.GetLogger().Tracef("no topic found for specify id:[%s]", id)
		return
	}

	delete(s.idTopicMapping, id)
	for topic, _ := range topicMap {
		delete(s.topicCallbackMapping, topic)
	}
	logger.GetLogger().Tracef("remove topic:[%v]", topicMap)
}

func (s *CallbackManager) Get(topic string) interfaces.WebSocketMessageCallback {
	s.lock.RLock()
	defer s.lock.RUnlock()

	cb := s.topicCallbackMapping[topic]
	if cb == nil {
		return nil
	}

	return cb.callback
}

type TopicManager struct {
	topicPrefix sync.Map
}

func NewTopicManager() *TopicManager {
	return &TopicManager{
		topicPrefix: sync.Map{},
	}
}

func (m *TopicManager) GetCallbackManager(topic string) *CallbackManager {
	parts := strings.Split(topic, ":")
	prefix := topic
	if len(parts) == 2 && parts[1] != "all" {
		prefix = parts[0]
	}

	ret, _ := m.topicPrefix.LoadOrStore(prefix, NewCallbackManager(topic))
	return ret.(*CallbackManager)
}

func (m *TopicManager) Range(f func(key, value any) bool) {
	m.topicPrefix.Range(f)
}
